src/toil/lib/ec2.py

```diff
@@ -42,6 +132,33 @@ def __init__(self, resource, to_state, state):
             "Expected state of %s to be '%s' but got '%s'" %
             (resource, to_state, state))
 
+def establish_boto3_session(region_name: Optional[str] = None) -> Session:
+    """
+    This is the One True Place where Boto3 sessions should be established, and
+    prepares them with the nexessary credential caching.
+
+    :param region_name: If given, the session will be associated with the given AWS region.
+    """
+
+    # Make sure to use credential caching when talking to Amazon via boto3
+    # See https://github.com/boto/botocore/pull/1338/
+    # And https://github.com/boto/botocore/commit/2dae76f52ae63db3304b5933730ea5efaaaf2bfc
+
+    botocore_session = get_session()
+    botocore_session.get_component('credential_provider').get_provider(
+        'assume-role').cache = JSONFileCache()
+    return Session(botocore_session=botocore_session, region_name=region_name)
```

Comment:
I guess we could try and read an environment variable here? But a boto3 session doesn't always need to be associated with a region, and sometimes you might want no region. We definitely can't ask the provisioner here since there might not be one.